toolchain/check/generic.cpp

```diff
@@ -312,83 +278,84 @@ static auto PopulateConstantsFromDeclaration(
-  for (auto i : llvm::seq(num_dependent_insts)) {
-    auto [inst_id, dep_kind] =
-        context.generic_region_stack().PeekDependentInsts()[i];
-
-    // If the type is symbolic, replace it with a type specific to this generic.
-    if ((dep_kind & GenericRegionStack::DependencyKind::SymbolicType) !=
-        GenericRegionStack::DependencyKind::None) {
-      auto inst = context.insts().Get(inst_id);
-      auto type_id = AddGenericTypeToEvalBlock(
-          context, generic_id, region, context.insts().GetLocId(inst_id),
-          constants_in_generic, inside_redeclaration, inst.type_id());
-      // If the generic declaration is invalid, it can result in an error.
-      if (type_id == SemIR::ErrorInst::TypeId) {
-        break;
-      }
-      // TODO: Eventually, completeness requirements should be modeled as
-      // constraints on the generic rather than properties of the type. For now,
-      // require the transformed type to be complete if the original was.
-      if (context.types().IsComplete(inst.type_id())) {
-        CompleteTypeOrCheckFail(context, type_id);
-      }
-      inst.SetType(type_id);
-      context.sem_ir().insts().Set(inst_id, inst);
-    }
+  context.generic_region_stack().AddDependentInst(dependent_inst.inst_id);
 
-    // If the instruction has a symbolic constant value, then make a note that
-    // we'll need to evaluate this instruction when forming the specific. Update
-    // the constant value of the instruction to refer to the result of that
-    // eventual evaluation.
-    if ((dep_kind & GenericRegionStack::DependencyKind::SymbolicConstant) !=
-        GenericRegionStack::DependencyKind::None) {
-      // Update the constant value to refer to this generic.
-      context.constant_values().Set(
-          inst_id, AddGenericConstantToEvalBlock(
-                       context, generic_id, region, constants_in_generic,
-                       inside_redeclaration, inst_id));
+  auto [generic_id, region] =
+      context.generic_region_stack().PeekPendingGeneric();
+  if (!generic_id.has_value()) {
+    // Allocate a placeholder generic now to form a generic ID. This generic
+    // will be populated once we reach the end of the generic declaration.
+    generic_id = context.generics().Add(
+        SemIR::Generic{.decl_id = SemIR::InstId::None,
+                       .bindings_id = SemIR::InstBlockId::None,
+                       .self_specific_id = SemIR::SpecificId::None});
+    context.generic_region_stack().SetPendingGenericId(generic_id);
+  }
+  auto& constants_in_generic =
+      context.generic_region_stack().PeekConstantsInGenericMap();
+
+  // If the type is symbolic, replace it with a type specific to this generic.
+  if ((dep_kind & DependentInst::SymbolicType) != DependentInst::None) {
+    auto inst = context.insts().Get(inst_id);
+    auto type_id = AddGenericTypeToEvalBlock(
+        context, generic_id, region, context.insts().GetLocId(inst_id),
```

Comment:
```suggestion
        context, generic_id, region, /*loc_id=*/inst_id,
```
We can pass `inst_id` as a location now

Reply:
I've tried that but it doesn't work:
```
FATAL failure at toolchain/sem_ir/absolute_node_id.cpp:81: Unexpected LocId: loc_inst44
```
so I've backed it out again. I guess this causes us to create an instruction whose location is another instruction, which is apparently disallowed but not enforced by the type system any more?

Reply:
Right, sorry, this breaks various things that consume LocId at the moment, and will be addressed by https://github.com/carbon-language/carbon-lang/pull/5355